bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java

```diff
@@ -1947,6 +1937,156 @@ protected void destroy() {
 	}
 }
 
+private abstract class ImageFromImageDataProviderWrapper extends AbstractImageProviderWrapper {
+
+	protected abstract ElementAtZoom<ImageData> loadImageData(int zoom);
+
+	void initImage() {
+		// As the init call configured some Image attributes (e.g. type)
+		// it must be called
+		ImageData imageDataAt100 = getImageData(100);
+		init(imageDataAt100, 100);
+		destroyHandleForZoom(100);
+	}
+
+	@Override
+	ImageData getImageData(int zoom) {
+		if (zoomLevelToImageHandle.containsKey(zoom)) {
+			return zoomLevelToImageHandle.get(zoom).getImageData();
+		}
+		if (!zoomLevelToImageHandle.isEmpty()) {
+			return getScaledImageData(zoom);
+		}
+		ElementAtZoom<ImageData> loadedImageData = loadImageData(zoom);
+		return DPIUtil.scaleImageData(device, loadedImageData, zoom);
+	}
+
+	@Override
+	ImageHandle getImageMetadata(int zoom) {
+		if (zoomLevelToImageHandle.containsKey(zoom)) {
+			return zoomLevelToImageHandle.get(zoom);
+		} else  {
+			ImageData scaledImageData = getImageData(zoom);
+			ImageHandle imageHandle = init(scaledImageData, zoom);
+			return imageHandle;
+		}
+	}
+}
+
+private class PlainImageDataProviderWrapper extends ImageFromImageDataProviderWrapper {
+	private ImageData imageDataAt100;
+
+	PlainImageDataProviderWrapper(ImageData imageData) {
+		this.imageDataAt100 = (ImageData) imageData.clone();
+		initImage();
+	}
+
+	@Override
+	protected Rectangle getBounds(int zoom) {
+		Rectangle rectangle = new Rectangle(0, 0, imageDataAt100.width, imageDataAt100.height);
+		return DPIUtil.scaleUp(rectangle, zoom);
+	}
+
+	@Override
+	protected ElementAtZoom<ImageData> loadImageData(int zoom) {
+		return new ElementAtZoom<>(imageDataAt100, 100);
+	}
+
+	@Override
+	AbstractImageProviderWrapper createCopy(Image image) {
+		return image.new PlainImageDataProviderWrapper(this.imageDataAt100);
+	}
+}
+
+private class MaskedImageDataProviderWrapper extends ImageFromImageDataProviderWrapper {
+	private final ImageData srcAt100;
+	private final ImageData maskAt100;
+
+	MaskedImageDataProviderWrapper(ImageData srcAt100, ImageData maskAt100) {
+		this.srcAt100 = (ImageData) srcAt100.clone();
+		this.maskAt100 = (ImageData) maskAt100.clone();
+		initImage();
+	}
+
+	@Override
+	protected Rectangle getBounds(int zoom) {
+		Rectangle rectangle = new Rectangle(0, 0, srcAt100.width, srcAt100.height);
+		return DPIUtil.scaleUp(rectangle, zoom);
+	}
+
+	@Override
+	protected ElementAtZoom<ImageData> loadImageData(int zoom) {
+		ImageData scaledSource = DPIUtil.scaleImageData(device, srcAt100, zoom, 100);
+		ImageData mask = ImageData.convertMask(maskAt100);
+		ImageData scaledMask = DPIUtil.scaleImageData(device, mask, zoom, 100);
```

Comment:
Two questions on this:
1. The order of these operations has changed. Is that correct?
2. If yes, can't we sotre the converted mask instead of cloning and storing the original one already in the constructor?

Reply:
Thanks for the clarification and correction!